functions/go/set-image/transformer/images_transformer.go

```diff
@@ -0,0 +1,196 @@
+	Digest string `json:"digest,omitempty" yaml:"digest,omitempty"`
+}
+
+// setImageResult keeps the old value and new value for logging
+type setImageResult struct {
+	currentValue  string
+	proposedValue string
+}
+
+// SetImage supports the set-image workflow, it uses Config to parse functionConfig, Transform to change the image
+type SetImage struct {
+	// Image is the desired image
+	Image Image `json:"image,omitempty" yaml:"image,omitempty"`
+	// ConfigMap keeps the data field that holds image information
+	DataFromDefaultConfig map[string]string `json:"data,omitempty" yaml:"data,omitempty"`
+	// ONLY for kustomize, AdditionalImageFields is the user supplied fieldspec
+	AdditionalImageFields types.FsSlice `json:"additionalImageFields,omitempty" yaml:"additionalImageFields,omitempty"`
+	// resultCount logs the total count image change
+	resultCount int
+}
+
+// Run implements the Runner interface that transforms the resource and log the results
+func (t SetImage) Run(ctx *fn.Context, functionConfig *fn.KubeObject, items fn.KubeObjects) {
+	err := t.configDefaultData()
+	if err != nil {
+		ctx.ResultErrAndDie(err.Error(), nil)
+	}
+	err = t.validateInput()
+	if err != nil {
+		ctx.ResultErrAndDie(err.Error(), nil)
+	}
+
+	for _, o := range items {
+		var result []setImageResult
+		switch o.GetKind() {
+		case "Pod":
+			err, result = t.setPodContainers(o)
+		case "Deployment", "StatefulSet", "ReplicaSet", "DaemonSet", "PodTemplate":
+			err, result = t.setPodSpecContainers(o)
+		}
+		t.logResult(ctx, err, result, o)
+	}
+
+	if t.AdditionalImageFields != nil {
+		custom.SetAdditionalFieldSpec(functionConfig.GetMap("image"), items, functionConfig.GetSlice("additionalImageFields"), ctx)
+	}
+
+	summary := fmt.Sprintf("summary: updated a total of %v image(s)", t.resultCount)
+	ctx.ResultInfo(summary, nil)
+}
+
+// configDefaultData transforms the data from ConfigMap to SetImage struct
+func (t *SetImage) configDefaultData() error {
+	for key, val := range t.DataFromDefaultConfig {
+		switch key {
+		case "name":
+			t.Image.Name = val
+		case "newName":
+			t.Image.NewName = val
+		case "newTag":
+			t.Image.NewTag = val
+		case "digest":
+			t.Image.Digest = val
+		default:
+			return fmt.Errorf("ConfigMap has wrong field name %v", key)
+		}
+	}
+	return nil
+}
+
+// validateInput validates the inputs passed into via the functionConfig
+func (t *SetImage) validateInput() error {
+	// TODO: support container name and only one argument input in the next PR
+	if t.Image.Name == "" {
+		return fmt.Errorf("must specify `name`")
+	}
+	if t.Image.NewName == "" && t.Image.NewTag == "" && t.Image.Digest == "" {
+		return fmt.Errorf("must specify one of `newName`, `newTag`, or `digest`")
+	}
+	return nil
+}
+
+// updateContainerImages updates the images inside containers, return potential error, and a list of logging results
+func (t *SetImage) updateContainerImages(pod *fn.SubObject) (error, []setImageResult) {
```

Comment:
nit: Normally in the return values, the error is placed at the end. e.g. `([]setImageResult, error)`